### PR TITLE
MueLu: various changes

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
- Unit test for Stratimikos smoother
- Improvements for reuse scenarios
- Fix FPE in cut-based distance Laplacian

### CODE REVIEW

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp
Comment (mayrmt) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp:842]: @cgcgcg I find this variable name a little confusing. To me, its name indicates that it will be `true` when something has been added to the diagonal. Yet, here it's set to `true` if `row!=col`, i.e. if something has been added to the off-diagonal.

**Review by mayrmt — APPROVED** [review]
@cgcgcg I made a small remark about a possibly missleading variabel name. Feel free to address it, if you agree.

Thanks for adding unit tests!

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp
Comment (mayrmt) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp:871]: I'm wondering about the purpose of this. What additional debugging information does it provide to the user?

Comment (cgcgcg) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp:871]: I think it was just meant to initialize to some easily identifiable value.

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp
Comment (cgcgcg) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp:842]: Things are being lumped from off-diagonal to diagonal..

**Comment by jhux2** [comment]
@cgcgcg So how Dirichlet bc's were being handled caused an FPE?

**Comment by cgcgcg** [comment]
Yep, the situation was that BCs were enforced on rows, but not columns, but actually in hard fashion, so no explicit zeros in those rows. This meant that nothing ever got lumped to the distance Laplacian diagonal, which was consequently zero. We then use L_ij^2/(L_ii * L_jj) for i,j in a BC column, and hit that FPE in L_jj.

**Comment by bathmatt** [comment]
This test failed...

Test Name: Trilinos_pullrequest_python_3

set retest, as that seems like a testing glitch

**Comment by bathmatt** [comment]
@jmgate can you snap this in tomorrow if it isn't resolved by then.
Thanks!

**Comment by bartlettroscoe** [comment]
CC: @bathmatt 

@trilinos/framework, this PR is is being blocked by the framework team test  `TrilinosFrameworkTests_PullRequestLinuxDriverTest_UnitTests` that is failing in the `python_3` build that has nothing to do with this PR.  Can someone see why this test is failing and fix this?

**Comment by jmgate** [comment]
@bartlettroscoe, #8643 should address the problem.

@bathmatt, I'm out on vacation till Tuesday.  @dc-snl, do you have time to handle it?  If not, any EM-Plasma Owner or Maintainer can push changes to our Trilinos fork.  Or we could just wait till next week.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe, #8643 should address the problem.

@jmgate, except that test `TrilinosFrameworkTests_PullRequestLinuxDriverTest_UnitTests` in the `python_3` build is failing in PR #8643 as well as shown [here](https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-8643-test&field2=buildstarttime&compare2=84&value2=NOW) and [here](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercount=3&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=buildname&compare2=65&value2=PR-8643-test&field3=status&compare3=61&value3=failed).  (And the fact there are other failing tests as well is scary.  I will post one for the randomly failing Teuchos test.)

**Pull request merged**
